junit-platform-commons/src/main/java/org/junit/platform/commons/support/Resource.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015-2024 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.commons.support;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.function.Predicate;
+
+import org.apiguardian.api.API;
+
+/**
+ * Represents a resource on the classpath.
+ *
+ * @see ReflectionSupport#findAllResourcesInClasspathRoot(URI, Predicate, Predicate)
+ * @see ReflectionSupport#findAllResourcesInPackage(String, Predicate, Predicate)
+ * @see ReflectionSupport#findAllResourcesInModule(String, Predicate, Predicate)
+ * @see ReflectionSupport#streamAllResourcesInClasspathRoot(URI, Predicate, Predicate)
+ * @see ReflectionSupport#streamAllResourcesInPackage(String, Predicate, Predicate)
+ * @see ReflectionSupport#streamAllResourcesInModule(String, Predicate, Predicate)
+ */
+@API(status = EXPERIMENTAL, since = "1.11")
+public interface Resource {
+
+	/**
+	 * Get the resource name.
+	 * <p>
+	 * The resource name is a {@code /}-separated path. The path is relative to
+	 * the classpath root in which the resource is located.
+	 *
+	 * @return the resource name; never {@code null}
+	 */
+	String getName();
```

Comment:
I have intentionally expanded the original definition of `Resource` somewhat. 

A few reasons:
 * To make it possible apply the package filter in the `EngineDiscoveryRequestResolver`, `ReflectionUtils`. needs a `Predicate<String> resourceNameFilter`. The `Predicate<Resource> resourceFilter` should have access to the at least as much information as the `Predicate<String>`.
  
 * While the URI does identify a resource uniquely, it is rooted in the file system. This makes it a less convenient as a human readable identifier. For example in Cucumber, feature files locations are identified by their classpath resource name  or the path relative to the current working directory. For example
 
 ```
 Scenario: Addition                    # io/cucumber/examples/calculator/basic_arithmetic.feature:7    <<<<<<<<<<<<
  Given a calculator I just turned on # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.a_calculator_I_just_turned_on()
  When I add 4 and 5                  # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.adding(int,int)
  Then the result is 9                # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.the_result_is(double)

Scenario Outline: Many additions      # io/cucumber/examples/calculator/basic_arithmetic.feature:30     <<<<<<<<<<<<
  Given a calculator I just turned on # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.a_calculator_I_just_turned_on()
  Given the previous entries:         # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.thePreviousEntries(java.util.List<io.cucumber.examples.calculator.RpnCalculatorStepDefinitions$Entry>)
    | first | second | operation |
    | 1     | 1      | +         |
    | 2     | 1      | +         |
  When I press +                      # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.I_press(java.lang.String)
  And I add 2 and 3                   # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.adding(int,int)
  And I press +                       # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.I_press(java.lang.String)
  Then the result is 10               # io.cucumber.examples.calculator.RpnCalculatorStepDefinitions.the_result_is(double)
```